jwst/ramp_fitting/ols_fit.py

```diff
@@ -0,0 +1,3491 @@
+    #  final group for this semiramp is not final group of the whole ramp)
+    wh_check = np.where((l_interval > 2) & (end_locs != ngroups - 1) & ~pixel_done)
+    if len(wh_check[0]) > 0:
+        f_max_seg = fit_next_segment_long_not_end_of_ramp(
+            wh_check, start, end_st, end_heads, pixel_done, got_case, f_max_seg,
+            inv_var, num_seg, slope, intercept, variance, sig_intercept, sig_slope,
+            opt_res, save_opt, mask_2d_init, end_locs, ngroups)
+
+    # CASE: interval too short to fit normally (only 2 good groups)
+    #    At end of array, NGROUPS>1, but exclude NGROUPS==2 datasets
+    #    as they are covered in `fit_short_ngroups`.
+    wh_check = np.where((l_interval == 1) & (end_locs == ngroups - 1)
+                        & (ngroups > 2) & (~pixel_done))
+
+    if len(wh_check[0]) > 0:
+        f_max_seg = fit_next_segment_short_seg_at_end(
+            wh_check, start, end_st, end_heads, pixel_done, got_case, f_max_seg,
+            inv_var, num_seg, slope, intercept, variance, sig_intercept, sig_slope,
+            opt_res, save_opt, mask_2d_init)
+
+    # CASE: full-length ramp has 2 good groups not at array end
+    wh_check = np.where((l_interval == 2) & (ngroups > 2)
+                        & (end_locs != ngroups - 1) & ~pixel_done)
+
+    if len(wh_check[0]) > 0:
+        f_max_seg = fit_next_segment_short_seg_not_at_end(
+            wh_check, start, end_st, end_heads, pixel_done, got_case, f_max_seg,
+            inv_var, num_seg, slope, intercept, variance, sig_intercept, sig_slope,
+            opt_res, save_opt, mask_2d_init, end_locs, ngroups)
+
+    # CASE: full-length ramp has a good group on 0th group of the entire ramp,
+    #    and no later good groups. Will use single good group data as the slope.
+    wh_check = np.where(
+        mask_2d_init[0, :] & ~mask_2d_init[1, :] & (ramp_mask_sum == 1) & ~pixel_done)
+
+    if len(wh_check[0]) > 0:
+        f_max_seg = fit_next_segment_only_good_0th_group(
+            wh_check, start, end_st, end_heads, pixel_done, got_case, f_max_seg,
+            inv_var, num_seg, slope, intercept, variance, sig_intercept, sig_slope,
+            opt_res, save_opt, mask_2d_init)
+
+    # CASE: the segment has a good 0th group and a bad 1st group.
+    wh_check = np.where(mask_2d_init[0, :] & ~mask_2d_init[1, :] & ~pixel_done
+                        & (end_locs == 1) & (start == 0))
+
+    if len(wh_check[0]) > 0:
+        fit_next_segment_good_0th_bad_1st(
+            wh_check, start, end_st, end_heads, got_case, ngroups)
+
+    # CASE OTHER: all other types of segments not covered earlier. No segments
+    #   handled here have adequate data, but the stack arrays are updated.
+    wh_check = np.asarray(np.where(~pixel_done & ~got_case))
+    if len(wh_check[0]) > 0:
+        fit_next_segment_all_other(wh_check, start, end_st, end_heads, ngroups)
+
+    return f_max_seg, num_seg
+
```

Comment:
You may want to separate out the related fit_next_XYZ ('case') functions into another file.

Reply:
I second the above suggestion.

Reply:
The `ols_fit.py` file is a large file and could be broken down into several smaller files.  Should other parts of the code be identified to separate into their own file?

Reply:
To further segregate related functions in ols_fit.py,  here are some rough groupings of similar functions that you could consider moving into other modules.  (This is just based on a quick review, and there is some overlap here  - consider this just as a guideline): 

# Output model-related 
def set_output_models( )
def create_output_models( )

# OLS ramp fitting
def ols_ramp_fit_multi( )
def ols_ramp_fit_sliced()
def ols_ramp_fit_single()

# JWST-specific
def discard_miri_groups( )

# Slope fitting
def ramp_fit_slopes( )
def ramp_fit_compute_variances( )
def ramp_fit_overall( ) # needs epydoc description
def calc_slope( )

# Perhaps move to utils.py
def calc_power(snr)
def interpolate_power(snr) # needs epydocs

# Fitting particular segments/ramps
def fit_next_segment( ) 
def fit_lines( )
def fit_single_read( )
def fit_double_read( )
def fit_1_group( )
def fit_2_group( )
def calc_num_seg( )

# Calculation of goodness-of-fit parameters
def calc_unwtd_fit( )
def calc_opt_fit( )
def calc_unwtd_sums( )
def calc_opt_sums( )